chess/ChessGameModel.swift

```diff
@@ -15,22 +16,88 @@ struct ChessGameModel {
     }
 
     private(set) var board: [[ChessPiece?]] = .init(repeating: .init(repeating: nil, count: 8), count: 8)
+    private let logger = Logger(subsystem: "com.khaletska.chess", category: "GameModel")
 
     mutating func createNewGameBoard(configuration: BoardConfiguration) {
         self.board = configuration.generateBoard()
     }
 
-    mutating func move(from: Coordinate, to: Coordinate) {
-        let piece = self.board[from.row][from.col]
-        self.board[to.row][to.col] = piece
-        self.board[from.row][from.col] = nil
+    mutating func pieceMoved(from source: Coordinate, to destination: Coordinate) throws {
+        guard let piece = self.board[source.row][source.col] else {
+            throw ChessGameModelError.missingPiece
+        }
+
+        if isPawnPromotionMove(piece: piece, to: destination) {
+            try self.promotePawn(from: source, to: destination)
+            self.logger.log("Pawn \(piece) moved from \(source) to \(destination) and promoted to \(ChessPiece(kind: .queen, color: piece.color))")
+        }
+        else {
+            movePiece(from: source, to: destination)
+            self.logger.log("Piece \(piece) moved from \(source) to \(destination)")
+        }
+    }
+
+    // MARK: - Actions -
+
+    private mutating func movePiece(from source: Coordinate, to destination: Coordinate) {
+        self.board[destination.row][destination.col] = self.board[source.row][source.col]
+        self.board[source.row][source.col] = nil
+    }
+
+    private mutating func promotePawn(from source: Coordinate, to destination: Coordinate) throws {
+        guard let piece = self.board[source.row][source.col] else {
+            throw ChessGameModelError.missingPiece
+        }
+
+        guard case .pawn = piece.kind else {
+            throw ChessGameModelError.invalidPromotion
+        }
+
+        guard (piece.color == .white && destination.row == 0) || (piece.color == .black && destination.row == 7) else {
+            throw ChessGameModelError.invalidPromotion
+        }
+
+        self.board[destination.row][destination.col] = ChessPiece(kind: .queen, color: piece.color)
+        self.board[source.row][source.col] = nil
+    }
+
+    // MARK: - Conditions -
+
+    private func isPawnPromotionMove(piece: ChessPiece, to: Coordinate) -> Bool {
+        guard piece.kind == .pawn else {
+            return false
+        }
+
+        guard (piece.color == .white && to.row == 0) || (piece.color == .black && to.row == 7) else {
+            return false
+        }
+
+        return true
     }
 
 }
 
-struct Coordinate: Equatable {
+struct Coordinate: Equatable, CustomStringConvertible {
 
     let row: Int
     let col: Int
 
+    var description: String {
```

Comment:
in case you are lazy to check debugger logs, this is how logs look like:
<img width="605" alt="image" src="https://github.com/user-attachments/assets/7612e435-33f0-4e3e-9068-57f6e65d4977">

Reply:
Are these pictures clear for you? So smol so I am not sure are readable

Reply:
But I am fine with them if you think they are readable enough

Reply:
they are small in terms of size (since they are part of font) but they are descriptive since font has both black and white pieces. We can use letter coding if this will be more logical. Intention was to bring the idea, but I'm fine with any implementation of this idea